File: clients/android/app/src/main/kotlin/com/lextures/android/core/i18n/LocaleApi.kt
package com.lextures.android.core.i18n

import com.lextures.android.core.network.ApiClient
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

object LocaleApi {
    @Serializable
    data class LocaleRequest(val locale: String)

    @Serializable
    data class LocaleResponse(val locale: String? = null)

    private val json = Json { ignoreUnknownKeys = true }

    suspend fun saveLocale(tag: String, accessToken: String, client: ApiClient = ApiClient()): String =
        withContext(Dispatchers.IO) {
            val body = json.encodeToString(LocaleRequest.serializer(), LocaleRequest(tag))
            val (responseBody, _) = client.request(
                path = "/api/v1/settings/locale",
                method = "PUT",
                body = body,
                accessToken = accessToken,
            )
            val response = json.decodeFromString(LocaleResponse.serializer(), responseBody)
            response.locale?.trim()?.takeIf { it.isNotEmpty() } ?: tag
        }
}

File: clients/ios/Lextures/Core/I18n/LocaleAPI.swift
import Foundation

enum LocaleAPI {
    struct LocaleResponse: Decodable {
        var locale: String?
    }

    struct LocaleRequest: Encodable {
        var locale: String
    }

    /// Persists the user's locale preference on the server (web parity).
    static func saveLocale(_ tag: String, accessToken: String) async throws -> String {
        let (data, _) = try await APIClient().request(
            path: "/api/v1/settings/locale",
            method: "PUT",
            body: LocaleRequest(locale: tag),
            authorized: true,
            accessToken: accessToken
        )
        let response = try JSONDecoder().decode(LocaleResponse.self, from: data)
        return response.locale?.trimmingCharacters(in: .whitespacesAndNewlines).nonEmpty ?? tag
    }
}

private extension String {
    var nonEmpty: String? {
        isEmpty ? nil : self
    }
}
